src/compose.ts

```diff
@@ -274,17 +285,99 @@ function normalizeArrayOfStrings(value: any[]): string[] {
 	return _.map(value, String);
 }
 
-function validateServiceVolume(serviceVolume: string, volumeNames: string[]) {
-	const colonIndex = serviceVolume.indexOf(':');
-	if (colonIndex === -1) {
-		throw new ValidationError(`Invalid volume: '${serviceVolume}'`);
+function normalizeServiceBuild(
+	serviceBuild: string | BuildConfig,
+	networkNames: string[],
+): BuildConfig {
+	if (typeof serviceBuild === 'string') {
+		serviceBuild = { context: serviceBuild };
+	}
+	if (serviceBuild.args) {
+		serviceBuild.args = normalizeKeyValuePairs(serviceBuild.args);
+	}
+	if (serviceBuild.labels) {
+		serviceBuild.labels = normalizeKeyValuePairs(serviceBuild.labels);
+		validateLabels(serviceBuild.labels);
+	}
+	if (serviceBuild.extra_hosts && !_.isArray(serviceBuild.extra_hosts)) {
+		serviceBuild.extra_hosts = normalizeExtraHostObject(
+			serviceBuild.extra_hosts as any,
+		);
+	}
+	if (serviceBuild.isolation) {
+		throw new ValidationError('service.build.isolation is not allowed');
+	}
+	if (
+		serviceBuild.network &&
+		serviceBuild.network !== 'host' &&
+		serviceBuild.network !== 'none'
+	) {
+		if (networkNames.indexOf(serviceBuild.network) === -1) {
+			throw new ValidationError(
+				`Missing network definition for '${serviceBuild.network}'`,
+			);
+		}
 	}
-	const source = serviceVolume.slice(0, colonIndex);
-	if (path.parse(source).dir !== '') {
-		throw new ValidationError('Bind mounts are not allowed');
+	return serviceBuild;
+}
+
+function normalizeServiceVolumes(
+	serviceVolumes: any[],
+	volumeNames: string[],
+): string[] {
+	return serviceVolumes.map((volume) => {
+		const ref = normalizeServiceVolume(volume);
+		validateServiceVolume(ref, volumeNames);
+		return `${ref.source}:${ref.target}${ref.read_only ? ':ro' : ''}`;
+	});
+}
+
+function normalizeServiceVolume(serviceVolume: string | VolumeRef): VolumeRef {
+	let ref: VolumeRef = { type: 'volume' };
+	if (typeof serviceVolume === 'string') {
+		const parts = serviceVolume.split(':');
+		if (parts.length < 2) {
+			throw new ValidationError(`Invalid volume: '${serviceVolume}'`);
+		}
+		ref.source = parts[0];
+		ref.target = parts[1];
+		if (path.parse(ref.source).dir) {
```

Comment:
@dfunckt this was changed from checking if the dir property is there, changed it to the truthy check after an earlier review

Reply:
Do you refer to [this comment from @karaxuna](https://github.com/balena-io-modules/resin-compose-parse/pull/59#discussion_r696732067)? How is checking whether a property exists semantically equivalent to checking whether it's the empty string? What does this `if` do in the first place?

Reply:
Specifically with regard to the truthy evaluation of the following two expressions:
1. `path.parse(ref.source).dir !== ''`
2. `path.parse(ref.source).dir`

dir | 1 | 2
--- | --- | ---
undefined | true | false
null | true | false
'' | false | false
'x' | true | true

Reply:
with the `path.parse(source).dir` I check if the source is a volume name, which would return an empty string. this is to determine if I'm dealing with a bind/volume mount

Reply:
https://nodejs.org/api/path.html#path_path_parse_path doesn't mention what the return values can be, so I would assume you can not get anything but a valid string

Reply:
The fact it's explicitly checking for strict equality to `''` is strange and I wonder if this is because of different kind of values I encountered while testing this back then. I think it's better to just leave it as it was before @robertgzr unless you're able to find when/if values such as `null` and `undefined` are possible. I personally don't think it's worth going deeper into this though and I'd just trust that the initial check is correct.

Regardless, if you could leave a comment close by describing the intent of that check it'd be great.